discovery/support/chaincode/support.go

```diff
@@ -35,20 +35,58 @@ func NewDiscoverySupport(ci MetadataRetriever) *DiscoverySupport {
 	return s
 }
 
-func (s *DiscoverySupport) PoliciesByChaincode(channel string, cc string, _ ...string) []policies.InquireablePolicy {
-	chaincodeData := s.ci.Metadata(channel, cc, false)
+func (s *DiscoverySupport) PoliciesByChaincode(channel string, cc string, collections ...string) []policies.InquireablePolicy {
+	chaincodeData := s.ci.Metadata(channel, cc, collections...)
 	if chaincodeData == nil {
-		logger.Info("Chaincode", cc, "wasn't found")
+		logger.Infof("Chaincode %s wasn't found", cc)
 		return nil
 	}
+
+	// chaincode policy
 	pol := &common2.SignaturePolicyEnvelope{}
 	if err := proto.Unmarshal(chaincodeData.Policy, pol); err != nil {
-		logger.Warning("Failed unmarshaling policy for chaincode", cc, ":", err)
+		logger.Warningf("Failed unmarshaling policy for chaincode '%s': %s", cc, err)
 		return nil
 	}
 	if len(pol.Identities) == 0 || pol.Rule == nil {
-		logger.Warningf("Invalid policy, either Identities(%v) or Rule(%v) are empty:", pol.Identities, pol.Rule)
+		logger.Warningf("Invalid policy, either Identities(%v) or Rule(%v) are empty", pol.Identities, pol.Rule)
 		return nil
 	}
-	return []policies.InquireablePolicy{inquire.NewInquireableSignaturePolicy(pol)}
+	chaincodePolicy := inquire.NewInquireableSignaturePolicy(pol)
+
+	inquireablePolicies := map[policies.InquireablePolicy]bool{}
+	uniqueInquireablePolicies := []policies.InquireablePolicy{}
+
+	// chaincodeData.CollectionPolicies will be nil when using legacy lifecycle (lscc)
+	// because collection endorsement policies are not supported in lscc
+	if chaincodeData.CollectionPolicies == nil || len(collections) == 0 {
+		return []policies.InquireablePolicy{chaincodePolicy}
+	}
+	// process any collection policies
+	for _, collectionName := range collections {
+		// add the collection policy if it exists otherwise default to the chaincode policy
+		if policy, collectionPolicyExists := chaincodeData.CollectionPolicies[collectionName]; collectionPolicyExists {
+			pol := &common2.SignaturePolicyEnvelope{}
+			if err := proto.Unmarshal(policy, pol); err != nil {
+				logger.Warningf("Failed unmarshaling collection policy for chaincode '%s' collection '%s': %s", cc, collectionName, err)
```

Comment:
This logging should be an error IMO

Reply:
Ack, will also change the 2 warnings for the chaincode policy further above